walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
+        Collection<String> mandatoryAttendees = request.getAttendees();
+        Collection<String> allAttendees = request.getAllAttendees();
+        long duration = request.getDuration();
+
+        Collection<Event> mandatoryAttendeeEvents = getEventsForAttendees(mandatoryAttendees, events);
+        ArrayList<TimeRange> mandatoryAttendeesSlots = getPossibleSlots(mandatoryAttendeeEvents, mandatoryAttendees, duration);
+
+        Collection<Event> allAttendeeEvents = getEventsForAttendees(allAttendees, events);
+        ArrayList<TimeRange> allAttendeeSlots = getPossibleSlots(allAttendeeEvents, allAttendees, duration);
+
+        if (allAttendeeSlots.size() == 0 && mandatoryAttendees.size() != 0) {
+            return mandatoryAttendeesSlots;
+        }
+        else {
+            return allAttendeeSlots;
+        }
+
+    }
+
+    public Collection<Event> getEventsForAttendees(Collection<String> attendees, Collection<Event> events) {
+        ArrayList<Event> filteredEvents = new ArrayList<>();;
+        for (Event event : events) {
+            Set<String> eventAttendees = event.getAttendees();
+            for (String attendee : attendees) {
+                if (eventAttendees.contains(attendee)) {
+                    filteredEvents.add(event);
+                    break;
+                }
+            }
+        }
+        return filteredEvents;
+    }
+
+    public ArrayList<TimeRange> getPossibleSlots(Collection<Event> events, Collection<String> attendees, long duration) {
+        ArrayList<TimeRange> slots = new ArrayList<TimeRange>();
+        int attendeeNumber = attendees.size();
+
+        if (attendeeNumber == 0) {
+            slots.add(TimeRange.WHOLE_DAY);
+            return slots;
+        }
+        else if (duration > TimeRange.WHOLE_DAY.duration()) {
+            return slots;
+        }
+        else {
+            int eventTimeStart = TimeRange.START_OF_DAY;
+            int eventTimeEnd = TimeRange.START_OF_DAY;
+            int latestEventEnd = TimeRange.START_OF_DAY;
+            int start = TimeRange.START_OF_DAY;
+            int end = TimeRange.END_OF_DAY;
+
+            for (Event event : events) {
+                if (event.getWhen() == TimeRange.WHOLE_DAY) {
+                    return new ArrayList<TimeRange>();
+                }
+                eventTimeStart = event.getWhen().start();
+                eventTimeEnd = event.getWhen().end();
+                Set<String> eventAttendees = event.getAttendees();
+                if (start == eventTimeStart && end == eventTimeEnd) {
+                    return slots;
+                }
+                else if (eventTimeStart == start && eventTimeEnd < end) {
+                    start = eventTimeEnd;
+                }
+                else if (eventTimeStart > start && eventTimeStart - start >= duration && eventTimeEnd >= end) {
+                    slots.add(TimeRange.fromStartEnd(start, eventTimeStart, false));
+                    return slots;
```

Comment:
It looks like this only works if events is sorted by time. There will need to be some sorting before this or some verification that it's sorted